torch/_refs/nn/functional/__init__.py

```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
+    target: TensorLikeType,
+    weight: Optional[TensorLikeType],
+    reduction: str,
+    ignore_index: int,
+) -> TensorLikeType:
+    if input.ndim == 3 or input.ndim > 4:
+        msg = "Expected input dimension to be either [1, 2, 4] but recieved {}."
+        raise ValueError(msg.format(input.ndim))
+
+    if input.ndim != 1 and input.shape[0] != target.shape[0]:
+        msg = "Expected input batch size ({}) to match target batch size ({})."
+        raise ValueError(msg.format(input.shape[0], target.shape[0]))
+
+    _check_reduction_value(reduction)
+
+    flat_target = torch.reshape(target, [-1])
+    ignore_classes_mask = torch.eq(flat_target, ignore_index)
+    ignore_class_weight = torch.scalar_tensor(0, dtype=input.dtype, device=input.device)
+    default_class_weight = torch.scalar_tensor(
+        1, dtype=input.dtype, device=input.device
+    )
+
+    # TODO: This check does not work with FakeTensor inputs
+    """
+    num_classes = input.shape[1] if input.ndim > 1 else input.shape[0]
+    valid_classes_mask = torch.logical_and((flat_target >= 0), (flat_target < num_classes))
+    if not torch.all(torch.logical_or(ignore_classes_mask, valid_classes_mask)):
+        print(target, num_classes, ignore_index)
+        raise ValueError("Target class is out-of-bounds and not ignore index")
+    """
+
+    if weight is None:
+        current_weight = torch.where(
+            ignore_classes_mask,
+            ignore_class_weight,
+            default_class_weight,
+        )
+    else:
+        current_weight = torch.where(
+            ignore_classes_mask,
+            ignore_class_weight.expand_as(flat_target),
+            weight[flat_target],
+        )
+
+    if input.ndim == 1:
+        loss = -input[target] * current_weight
+    elif input.ndim == 2:
+        batch_size = input.shape[0]
+        loss = -input[torch.arange(batch_size), target] * current_weight
+    else:
+        batch_size = input.shape[0]
+        height = input.shape[2]
+        width = input.shape[3]
+        extent = height * width
+        numel = batch_size * extent
+        bdx = torch.arange(numel) // extent
+        hdx = (torch.arange(numel) - (bdx * extent)) // width
+        wdx = torch.arange(numel) % width
+        loss = -input[bdx, flat_target, hdx, wdx] * current_weight
+    loss = torch.reshape(loss, target.shape)
+
+    if reduction == "none":
+        return loss
+    elif reduction == "sum":
+        return torch.sum(loss)
+    else:
+        return torch.sum(loss) / torch.sum(current_weight)
+
+
+@register_decomposition(torch.ops.aten.nll_loss)
+def nll_loss(
+    input: TensorLikeType,
+    target: TensorLikeType,
+    weight: Optional[TensorLikeType] = None,
+    size_average: Optional[bool] = None,
+    ignore_index: int = -100,
+    reduce: Optional[bool] = None,
+    reduction: str = "mean",
+) -> TensorLikeType:
+    if size_average is not None or reduce is not None:
+        # TODO: raise exception instead of converting value
```

Comment:
What's the current state of the actual `nll_loss` operation?

Reply:
Currently it returns `UserWarning: size_average and reduce args will be deprecated, please use reduction='mean' instead.` when `reduce=True`

Reply:
Let's include that in this comment so a reader understands this behavior is for consistency with the current eager operation